### PR TITLE
Feat: Enable APNS registration + notification delivery in macOS apps

### PR DESCRIPTION
**Pull request by joanx**
#### Description of Change

Want to pick up on previous attempts https://github.com/electron/electron/pull/13777 and https://github.com/electron/electron/pull/18286 by @frantic and @kevinhu88 to enable APNS in Electron.

My understanding is that @miniak 's https://github.com/electron/electron/pull/18286 was blocked on concerns over supporting `UserNotification` and `NSUserNotifications` at the same time. This subset (APNS registration + notification receipt) doesn't actually seem to reference the `UserNotification` object (the relevant `UserNotification` -related code appears to have been handled in https://github.com/electron/electron/pull/25950.)

I've ported over the remaining functionality to register + receive APNS notifications into a separate fork (as the previous PR appears to be blocked on merge conflicts) and have confirmed successful registration + notification receipt via APNs with [this](https://github.com/electron/electron-quick-start/compare/master...joanx:master) sample app (note, I needed to create a provisioning profile + code-sign this app for successful APNS registration).

Would love to re-start the conversation around what it would take to enable APNS in Electron. Thanks in advance!

#### Checklist

- [x] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: 

Notes: Added support for push notifications from APNs for macOS apps.

### CODE REVIEW

File: shell/browser/browser_mac.mm [hunk 1]
```diff
@@ -532,4 +532,34 @@ void RemoveFromLoginItems() {
   }
 }
 
+void Browser::RegisterForRemoteNotifications() {
+  [[AtomApplication sharedApplication]
+      registerForRemoteNotificationTypes:NSRemoteNotificationTypeBadge |
```
Comment (joanx) [inline shell/browser/browser_mac.mm:537]: Attempting to use the updated API ([registerForRemoteNotifications](https://developer.apple.com/documentation/uikit/uiapplication/1623078-registerforremotenotifications?language=objc)) resulted in this error:

```
'registerForRemoteNotifications' has been marked as being introduced in macOS 10.14 here, but the deployment target is macOS 10.11.0
```

**Review by zcbenz — APPROVED** [review]
API LGTM

**Comment by joanx** [comment]
@zcbenz thanks so much for the review! Is there any guidance on next steps to get this merged?

**Comment by zcbenz** [comment]
We just need to wait for a few more reviews from maintainers.

**Comment by nornagon** [comment]
cc @miniak and @MarshallOfSound who have worked in this area previously

**Comment by michal-bednarz** [comment]
Hey 👋  @nornagon @zcbenz @miniak @MarshallOfSound I just wanted to kindly ping you about this PR :) What are the next steps, can we help somehow with pushing it into the release version?

**Comment by MarshallOfSound** [comment]
Haven't had a chance to look this over yet but I don't see CI runs this PR yet.  @joanx can you try signing in to circleci and pushing a new commit to this PR.  Builds should kick off after that

**Comment by joanx** [comment]
@MarshallOfSound just did!

**Comment by MarshallOfSound** [comment]
Few things:
* Code looks fine in principle, the lint is unhappy and should be fixed
* Idea is fine as well, I'm relatively OK with this API being added as it fundamentally has to be added into Electron if someone wants to use APNS
* This is the part where I bikeshed over where to put the API:

I'm not a fan of lumping everything in with `app.*`, it's kinda a catch-all that is incredibly oversaturated at this point.  I have two suggestions:

* `import { apns } from 'electron'`, specific to APNS
* `import { pushNotifications } from 'electron'` if we think we may implement the windows equivalent
* The existing `Notification` module, maybe `Notification.push` or `Notification.apns`

I'm open to arguments for either or an alternative, but I'm -0.5 on lumping this into `app`

**Comment by joanx** [comment]
@MarshallOfSound , thanks for the review. I'm unfortunately not familiar with the code organization and languages used here and unable to achieve a working binary with the refactor though I'd hugely appreciate any guidance on the in-progress changes.

Wondering if you have thoughts on any of these options:

1. Keep logic in `app`
2. Possible for me to pair with someone in the Electron maintainers group?
3. Possible for someone in the Electron maintainers group to own the refactor?
4. I continue to own these changes; there may be somewhat long iteration cycles as I ramp up on the necessary context to get this to a merge-ready state

Thank you!

**Review by nornagon — CHANGES_REQUESTED** [review]
marking request changes because the tests are legitimately failing!

**Comment by joanx** [comment]
Successfully tested APNS registration using [this](https://github.com/electron/electron-quick-start/compare/master...joanx:master) sample app post-refactor

**Review by nornagon — CHANGES_REQUESTED** [review]
Please remove the added `LICENSES.chromium.html` file, it shouldn't be checked in!

File: shell/browser/browser_mac.mm [hunk 2]
```diff
@@ -542,4 +542,33 @@ void RemoveFromLoginItems() {
...
+void Browser::RegisterForAPNSNotifications() {
+  [[AtomApplication sharedApplication]
+      registerForRemoteNotificationTypes:NSRemoteNotificationTypeBadge |
+                                         NSRemoteNotificationTypeAlert |
+                                         NSRemoteNotificationTypeSound];
+}
+
+void Browser::UnregisterForAPNSNotifications() {
+  [[AtomApplication sharedApplication] unregisterForRemoteNotifications];
+}
```
Comment (nornagon) [inline shell/browser/browser_mac.mm:554]: I don't think there's any reason for these methods to be on `Browser`, you should be able to move them to `PushNotifications` without any issues.

**Comment by joanx** [comment]
> Please remove the added `LICENSES.chromium.html` file, it shouldn't be checked in!

Done!

File: shell/browser/mac/electron_application_delegate.mm [hunk 3]
```diff
@@ -157,4 +157,31 @@ - (IBAction)newWindowForTab:(id)sender {
...
+- (void)application:(NSApplication*)application
+    didRegisterForRemoteNotificationsWithDeviceToken:(NSData*)deviceToken {
+  // https://stackoverflow.com/a/16411517
+  const char* tokenData = static_cast<const char*>([deviceToken bytes]);
+  NSMutableString* tokenString = [NSMutableString string];
+  for (NSUInteger i = 0; i < [deviceToken length]; i++) {
+    [tokenString appendFormat:@"%02.2hhX", tokenData[i]];
+  }
+  electron::Browser::Get()->DidRegisterForAPNSNotificationsWithDeviceToken(
+      base::SysNSStringToUTF8(tokenString));
```
Comment (nornagon) [inline shell/browser/mac/electron_application_delegate.mm:169]: Instead of directing this through `electron::Browser`, expose the `PushNotifications` singleton and access it directly. i.e.
```suggestion
  electron::api::PushNotifications::Get()->DidRegisterForAPNSNotificationsWithDeviceToken(
      base::SysNSStringToUTF8(tokenString));
```

along with the associated machinery for `PushNotifications` to be a singleton with a static `Get()` method.

There are a couple of things to consider here. In particular, the `PushNotifications` singleton is currently not created until the Node environment boots up (during the `Initialize` function), whereas `Browser` is always available. We need to avoid calling functions on an uninstantiated object, so I'd recommend:

1. When the `Initialize` function runs and the `PushNotifications` object is created, store a pointer to it in a global.
2. When the `PushNotifications` object is deleted, clear the global (set it to `nullptr`).
3. Create a new `PushNotifications::Get()` method which returns the (possibly null) value of the pointer.
4. When we access it here, first check that it's non-null.

This should also let you remove the new `BrowserObserver` methods.

Basically, we can do all this without involving the `Browser` object!

**Review by nornagon — CHANGES_REQUESTED** [review]

**Comment by joanx** [comment]
@tiagoporto there's still a bit of outstanding code review I'm hoping to get to shortly!

File: shell/browser/api/electron_api_push_notifications.cc [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+PushNotifications::~PushNotifications() {
+  DCHECK(push_notifications);
+  push_notifications = nullptr;
+}
+
+// static
+PushNotifications* PushNotifications::Get() {
+  return push_notifications;
+}
```
Comment (nornagon) [inline shell/browser/api/electron_api_push_notifications.cc:34]:
```suggestion
PushNotifications::PushNotifications() = default;

PushNotifications::~PushNotifications() {
  g_push_notifications = nullptr;
}

// static
PushNotifications* PushNotifications::Get() {
  if (!g_push_notifications)
    g_push_notifications = new PushNotifications();
  return g_push_notifications;
}
```

File: shell/browser/api/electron_api_push_notifications.cc [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+// static
+gin::Handle<PushNotifications> PushNotifications::Create(v8::Isolate* isolate) {
+  return gin::CreateHandle(isolate, new PushNotifications());
```
Comment (nornagon) [inline shell/browser/api/electron_api_push_notifications.cc:38]:
```suggestion
  return gin::CreateHandle(isolate, PushNotifications::Get());
```

Comment (nornagon) [reply shell/browser/api/electron_api_push_notifications.cc:38]: This change & the previous change together just add a little more safety, as the `PushNotifications::Create` function is public and could potentially be called multiple times. With these changes, calling `PushNotifications::Create` multiple times will just return multiple handles to the same object, which is fine.

File: shell/browser/api/electron_api_push_notifications_mac.mm [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+#import "shell/browser/mac/electron_application.h"
+#include "shell/common/gin_converters/value_converter.h"
+
+namespace electron {
+
+namespace api {
+
+void PushNotifications::RegisterForAPNSNotifications() {
+  [[AtomApplication sharedApplication]
+      registerForRemoteNotificationTypes:NSRemoteNotificationTypeBadge |
```
Comment (nornagon) [inline shell/browser/api/electron_api_push_notifications_mac.mm:18]: This method appears to be [deprecated](https://developer.apple.com/documentation/uikit/uiapplication/1623010-registerforremotenotificationtyp), can we use the [newer version](https://developer.apple.com/documentation/uikit/uiapplication/1623078-registerforremotenotifications?language=objc) instead?

File: shell/browser/mac/electron_application_delegate.mm [hunk 7]
```diff
@@ -157,4 +158,44 @@ - (IBAction)newWindowForTab:(id)sender {
   electron::Browser::Get()->NewWindowForTab();
 }
 
+- (void)application:(NSApplication*)application
+    didRegisterForRemoteNotificationsWithDeviceToken:(NSData*)deviceToken {
+  // https://stackoverflow.com/a/16411517
+  const char* tokenData = static_cast<const char*>([deviceToken bytes]);
```
Comment (nornagon) [inline shell/browser/mac/electron_application_delegate.mm:164]: code style for local variables is to use `snake_case`. it's a bit confusing when in Objective C land because argument names use `camelCase`! but the other functions in this file use `snake_case` for local variables so let's match suit.

File: docs/api/push-notifications.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+See: https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428430-application?language=objc
+
+## Methods
+
+The `pushNotification` module has the following methods:
+
+### `pushNotifications.registerForAPNSNotifications()` _macOS_
+
+Registers the app with Apple Push Notification service (APNS) to receive [Badge, Sound, and Alert](https://developer.apple.com/documentation/appkit/sremotenotificationtype?language=objc) notifications. If registration is successful, the BrowserWindow event `registered-for-apns-notifications` will be emitted with the APNS device token. Otherwise, the event `failed-to-register-for-apns-notifications` will be emitted.
+See: https://developer.apple.com/documentation/appkit/nsapplication/1428476-registerforremotenotificationtyp?language=objc
```
Comment (nornagon) [inline docs/api/push-notifications.md:45]: It seems like this might be better expressed as an asynchronous function (i.e. a Promise)?

i.e.

```suggestion
### `pushNotifications.registerForAPNSNotifications()` _macOS_

Returns `Promise<string>`

Registers the app with Apple Push Notification service (APNS) to receive [Badge, Sound, and Alert](https://developer.apple.com/documentation/appkit/sremotenotificationtype?language=objc) notifications. If registration is successful, the promise will be resolved with the APNS device token. Otherwise, the promise will be rejected with an error message.

See: https://developer.apple.com/documentation/appkit/nsapplication/1428476-registerforremotenotificationtyp?language=objc
```

File: docs/api/push-notifications.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
+# pushNotifications
+
+> Register for remote push notification services
+
+Process: [Main](../glossary.md#main-process)
```
Comment (nornagon) [inline docs/api/push-notifications.md:5]: I think this document would benefit from a short example explaining how these methods and events fit together.

**Review by nornagon — CHANGES_REQUESTED** [review]
Overall looking pretty good! Now that I'm looking more closely at the API, I think the `registerForAPNSNotifications` function might be better expressed as a Promise, rather than as a function which returns void and triggers one of two events later on. It might be a little trickier to implement in the C++ side, though.

File: shell/browser/api/electron_api_push_notifications_mac.mm (hunk 6 above)
Comment (joanx) [reply shell/browser/api/electron_api_push_notifications_mac.mm:29]: When I attempt to use `registerForRemoteNotifications` instead, I see:
```
../../electron/shell/browser/api/electron_api_push_notifications_mac.mm:28:40: error: 'registerForRemoteNotifications' is only available on macOS 10.14 or newer [-Werror,-Wunguarded-availability-new]
  [[AtomApplication sharedApplication] registerForRemoteNotifications];
                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sdk/xcode_links/MacOSX12.3.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSApplication.h:501:1: note: 'registerForRemoteNotifications' has been marked as being introduced in macOS 10.14 here, but the deployment target is macOS 10.13.0
- (void)registerForRemoteNotifications API_AVAILABLE(macos(10.14));
^
../../electron/shell/browser/api/electron_api_push_notifications_mac.mm:28:40: note: enclose 'registerForRemoteNotifications' in an @available check to silence this warning
  [[AtomApplication sharedApplication] registerForRemoteNotifications];
                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by joanx** [comment]
Updated to promisify APNS registration method, and tested [here](https://github.com/electron/electron-quick-start/compare/master...joanx:master#diff-27653c212e1cfe533e4eb2f7d0d3f89604c9de48a09583b4cbbbcbd08a07da79R40)

File: shell/browser/api/electron_api_push_notifications_mac.mm [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  apns_promise_set_.emplace_back(std::move(promise));
+  return handle;
+}
+
+void PushNotifications::ResolveAPNSPromiseSetWithToken(
+    const std::string& token_string) {
+  for (auto& promise : apns_promise_set_) {
+    promise.Resolve(token_string);
+  }
+  apns_promise_set_.clear();
```
Comment (nornagon) [inline shell/browser/api/electron_api_push_notifications_mac.mm:42]: Mega edge-case nitpick here, but to be on the safe side, it's best to make sure lists aren't modified during iteration. As a bonus, we can simplify this code a little:

```suggestion
  std::vector<gin_helper::Promise<std::string>> promises = std::move(apns_promise_set_);
  for (auto& promise : promises)
    promise.Resolve(token_string);
```

`std::move` on a vector clears out the original vector. If `register` is called again while we're in the middle of resolving promises, it won't affect the iteration, since we're operating on a separate vector now.

Same for the error method below :)

File: shell/browser/api/electron_api_push_notifications.h [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  const char* GetTypeName() override;
+
+  // disable copy
+  PushNotifications(const PushNotifications&) = delete;
+  PushNotifications& operator=(const PushNotifications&) = delete;
+
+#if BUILDFLAG(IS_MAC)
+  void OnDidReceiveAPNSNotification(const base::DictionaryValue& user_info);
+  void ResolveAPNSPromiseSetWithToken(const std::string& token_string);
+  void ResolveAPNSPromiseSetWithError(const std::string& error_message);
```
Comment (nornagon) [inline shell/browser/api/electron_api_push_notifications.h:43]: Nit:
```suggestion
  void RejectAPNSPromiseSetWithError(const std::string& error_message);
```

**Review by nornagon — CHANGES_REQUESTED** [review]
awesome awesome. did you happen to check how macOS behaves if you call `registerForAPNSNotifications()` twice?

e.g. does this work with the new code?

```js
const token1 = await pushNotifications.registerForAPNSNotifications()
const token2 = await pushNotifications.registerForAPNSNotifications()
```

does the 2nd promise properly resolve (or reject)? i.e. is it not left hanging forever? Also, is the token returned the same?

[edit] confirmed out-of-band that the above works as expected :)

**Comment by joanx** [comment]
Updated + re-tested!

File: shell/browser/api/electron_api_push_notifications_mac.mm [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+#include "shell/common/gin_converters/value_converter.h"
+#include "shell/common/gin_helper/promise.h"
+
+namespace electron {
+
+namespace api {
+
+// This set maintains all the promises that should be fulfilled
+// once macOS registers, or fails to register, for APNS
+std::vector<gin_helper::Promise<std::string>> apns_promise_set_;
```
Comment (nornagon) [inline shell/browser/api/electron_api_push_notifications_mac.mm:21]: ah missed this, this should be moved inside the class please!

Comment (joanx) [reply shell/browser/api/electron_api_push_notifications_mac.mm]: Updated!

**Review by nornagon — APPROVED** [review]

**Comment by tiagoporto** [comment]
Great job @joanx , looking forward to this merge.

**Comment by KishanBagaria** [comment]
Could this be backported to v20?

**Comment by erickzhao** [comment]
@KishanBagaria Unfortunately, we generally don't backport features to older branches at this point of the release cycle.

For more information on the process: https://github.com/electron/governance/blob/main/wg-releases/feature-backport-requests.md

**Comment by KishanBagaria** [comment]
@joanx out in https://github.com/electron/electron/releases/tag/v21.0.0

**Comment by JCBsystem** [comment]
@joanx  great work!

Is it possible to get the more detailed doc on this ? 
whats needed to use it?  entitlements / certs / signing .. mas build ... 
im stuck on how to get it in to my mas build.
thank you

**Comment by sumersao** [comment]
I have the same question ^ what entitlements do I need to add to get this to work?

**Comment by jarvis-zgp** [comment]
Failed to register for APNS: [Error: 13 NSOSStatusErrorDomain {
    NSDebugDescription = "Bundle identifier mismatch";
}]

May I ask where to configure the appId?
